### PR TITLE
Keep material and folder updater and update timestamp

### PR DESCRIPTION
**Pull request by fonglh**
Should the creator and created_at be preserved as well?

### CODE REVIEW

**Comment by weiqingtoh** [comment]
Would it make sense to abstract the duplicating flag to the duplication framework?

**Comment by weiqingtoh** [comment]
I feel that `created_by` and `updated_by` are system variables, and by convention, the original method is correct. 

Our assumption here is that since we display the `updater`, we assume that the `updaer` is also the `uploader`. This is technically true until we start duplicating things. If that is the case, it might be better for us to remove this assumption, and have another 2 columns - `uploader` and `uploaded_at` for materials and folder. 

Thoughts? @allenwq @fonglh

**Comment by allenwq** [comment]
It's the trade of between the consistent convention and maintaining the updating of two extra attributes. 

Also `uploader` might fail to work in the case when user only changes the material title or description, the attached file might not be replaced.  so `uploader` cannot totally replace `updater`.  But if we don't care about this case, I think having a `uploader` is a good idea.

**Comment by allenwq** [comment]
> Would it make sense to abstract the duplicating flag to the duplication framework?

Agree that it would be nice if the duplication flag could be set and cleared automatically during duplication, otherwise need to do quite a few models manually.

File: spec/services/course/duplication_service_spec.rb [hunk 1]
```diff
@@ -187,21 +187,35 @@
         let!(:folders) { create_list(:course_material_folder, 3, course: course) }
         let!(:content) { create(:course_material, folder: folders[0]) }
 
-        it 'duplicates the folders and their contents' do
+        before(:example) do
```
Comment (allenwq) [inline spec/services/course/duplication_service_spec.rb]: No need `:example`, just `before {}`

**Comment by fonglh** [comment]
I'll refactor the `@duplicating` flag in a future PR. That'll require more extensive testing as it touches quite a lot of the models.

File: app/models/course/material.rb [hunk 2]
```diff
@@ -1,21 +1,35 @@
...
+  belongs_to :folder, inverse_of: :materials, class_name: Course::Material::Folder.name
+
+  before_save :touch_folder
+  after_save :clear_duplication_flag
 
   validate :validate_name_is_unique_among_folders
   validates_with FilenameValidator
 
+  def touch_folder
+    folder.touch unless @duplicating
```
Comment (allenwq) [inline app/models/course/material.rb]: `folder.touch if !@duplicating && changed?`

**Comment by allenwq** [comment]
LGTM.

@weiqingtoh let's sit down and discuss this and #1720 on Mon. For now I will merge it so that @fonglh can test with duplication.

**Pull request merged**